Lab06 - Serial Peripheral Interface (#93)
UPB-FILS-MA/upb-fils-ma.github.io: docs/lab/06/index.md

---
sidebar_position: 7
description: Serial Peripheral Interface
slug: /lab/06
---

# 06 - Serial Peripheral Interface

This lab will teach you how to communicate with hardware devices using the Serial Peripheral Interface (SPI) protocol, in Embassy. 

## Resources

TBD

## Serial Peripheral Interface (SPI)

The Serial Peripheral Interface (or SPI) is a communication protocol used between microcontrollers and peripheral ICs (integrated circuits). It allows synchronous transfer of data between one *main* device and one or more *sub* devices (e.g. sensors). The main device is the one that dictates the communication.

### Data transmission

![SPI_protocol](images/spi_network.svg)

> During SPI transmission, there are 4 wires used:

Reviewer: Not necessarily, elaborate about half-duplex, full-duplex and transmission without CS.

Author: I explained that there is also a 3 wire SPI configuration, that provides a half-duplex communication. Anything else I should add?

- **CLK** - clock - generated by the main device - subs use it to synchronously read and write to the bus
- **MOSI** - **M**ain **O**ut **S**ub **I**n - used to carry data from the main device to *all* the subs
- **MISO** - **M**ain **I**n **S**ub **O**ut - used to carry data from the *active* sub to the main device
- **CS** - chip select - selects the sub to be activated, or the sub that the main wants to communicate with
                       - one wire per sub
                       - inactive subs will disconnect from the *MOSI* and *MISO* lines

:::info

> The communication is *full-duplex*. This means that, during data transmission, both the master and the slave **must write** to the MOSI/MISO wires, no matter if that data is relevant or not.

Reviewer 2: ```suggestion
The communication is *full-duplex*. This means that, during data transmission, both the main and the sub **must write** to the MOSI/MISO wires, no matter if that data is relevant or not.
```

> The communication is *full-duplex*. This means that, during data transmission, both the master and the slave **must write** to the MOSI/MISO wires, no matter if that data is relevant or not.

Reviewer: explain why... for every clk pulse dictated by main, 1 bit must be written

Author: Done. Is it clear now?

:::

#### Transmission example

> 1. main sets `CS` to `LOW` - the sub that the main wants to communicate with is activated - all other subs disconnect from `MOSI` and `MISO` lines

Reviewer 2: ```suggestion
1. main sets `CS` to `LOW` - the sub that the main wants to communicate with is activated - all other subs are already disconnect from `MOSI` and `MISO` lines as their CS line is not active
```

2. main writes the first bit on the `MOSI` line, and, *simultaneously*, sub writes the first bit on the `MISO` line
3. main starts the clock
4. on *rising edge* 
    - main reads data from the `MISO` line 
    - sub reads data from the `MOSI` line
5. on *falling edge*
    - main writes data to the `MOSI` line
    - sub writes data to the `MISO` line
6. repeat steps 4 and 5 until main stops the clock

:::note

> Whether steps 4 and 5 happen on rising/falling edge depends on the configuration of the SPI.

Reviewer 2: ```suggestion
Whether steps 4 and 5 happen on rising/falling edge depends on the `CPOL` configuration of the SPI.
```

:::

![SPI_example](images/spi_network_transmission.svg)

In the image above, the main first exchages data with the first sub by setting its corresponding `CS` line to low. After it finishes transmission with sub 1, it deactivates it and, afterwards, initiates communication with the second sub by setting its `CS` line to low. Notice that both main and sub read and write to the MOSI/MISO line simultaneously.

### SPI modes

> SPI has 4 different modes which define when data is read or written. These modes can be configured through 2 registers: `CPOL` and `CPHA`.

Reviewer: For this section please take a look at https://www.analog.com/en/resources/analog-dialogue/articles/introduction-to-spi-interface.html at Clock Polarity and Clock Phase

Author: Fixed it.

| Mode | `CPOL` | `CPHA` |
|-|-|-|
| 0 | 0 | 0 |
| 1 | 0 | 1 |
| 2 | 1 | 0 |
| 3 | 1 | 1 |

| | | |
|-|-|-| 
| `CPOL` | Clock polarity | defines when the data bit is read <br/> 0: *rising edge* <br/> 1: *falling edge* |
| `CPHA` | Clock phase | defines when the data is written to the line <br/> 0: when `CS` *activates* <br/> 1: on *clock edge* (depends on `CPOL`) |

### Daisy Chaining

> An SPI network with many sub devices can prove impractical from a hardware point of view. **Daisy chaining** is a method that solves this by linking multiple devices together with the following configuration:

Reviewer 2: This does not solve the problem of many subs, it just provides another way of linking subs. Subs here usually send in MISO the previous MOSI bit.

Author: Removed the part about hardware impracticality and provided a clearer explanation of the way daisy chaining works. Let me know if I should add anything else here.

![SPI_Daisy](images/spi_daisy.svg)

The `MOSI` line connects the main device only to the first sub. Then, the `MISO` output of the first sub ties to the `MOSI` input of the second sub, and so forth. The final sub in the chain ties to the `MISO` line to the main device. All `CS` lines are common, therefore all subs are active at the same time.

If main wants to communicate with the second sub, it will send its data through the first sub, and receive data from the second sub through all subs following it. (TODO: explain)

## DMA

When it comes to standard communication protocols that rely on [MMIO](/docs/lab/02), it would be inefficient to rely on the MCU itself to handle all of these data transfers. This is why the **Direct Memory Access** (DMA) is used. Its purpose is to offload the MCU by dealing with simple transmission to and from peripherals, and whenever it finishes a transfer, it raises an interrupt.

## SPI in Embassy

> In Embassy, we can use the SPI both blocking and asynchronously. The following example will be using the asynchronous version.

Reviewer 2: Add a warning explaining why the synchronous versions is not good.

Author: Done.

spi.transfer(&mut rx_buf, &tx_buf).await;
```

## Exercises

1. Connect the BMP280. Use Kicad to draw the schematic. (**1p**)
2. The example provided for exercise 2 in the lab skeleton is a base example of how to read a register of the BMP280. Modify it to read the `id` of the BMP280 and print it over serial. (**1p**)

:::tip
Use the datasheet to find the address of the `id` register! Take a look at section 4.2 and 4.3.
:::

3. Get the pressure and temperature readings from the sensor.
- Write the `ctrl_meas` register with appropiate configuration values. You can find information on the contents you should write to this register at section 4.3.4 of the datasheet. (**2p**)
- Read the pressure value and print it over the serial. (**1p**)

:::tip
The pressure value is split into 3 registers: `press_msb`, `press_lsb` and `press_xlsb`. 
- `press_msb` is the first binary half of the pressure value (stands for pressure most significant bits)
- `press_lsb` is the second binary half of the pressure value (stands for pressure least significant bits)
- `press_xlsb` is an extra degree of precision for the pressure value; **we won't be using it**

> To compute the pressure value, we need to read `press_msb` and `press_lsb`, and **add them together**.

Reviewer 2: ```suggestion
To compute the pressure value, we need to read `press_msb` and `press_lsb`, and **shift and add them together**.
```

> *pressure = `press_msb` + `press_lsb`*

Reviewer 2: ```suggestion
*pressure = `press_msb << 8` + `press_lsb`*
```

| 010 | oversampling x 2 |
| 011 | oversampling x 4 |
| 100 | oversampling x 8 |
| 101, others | oversampling x 16 |

| osrs_t[2:0] | Temperature oversampling |
|-|-|
| 000 | Skipped (output set to 0x80000) |
| 001 | oversampling x 1 |
| 010 | oversampling x 2 |
| 011 | oversampling x 4 |
| 100 | oversampling x 8 |
| 101, others | oversampling x 16 |

`press` registers - `press_msb`, `press_lsb`, `press_xslb` - contain the pressure measurement

`temp` registers - `temp_msb`, `temp_lsb`, `temp_xslb` - contain the temperature measurement

:::warning
Unless we write to the `osrs_p` and `osrs_t` fields of the `ctrl_meas` register, the `press` and `temp` registers will have a constant value of `0x80000`! We need to configure the pressure and temperature oversampling before reading the measurements.
:::

### BMP280 wiring

The BMP280 has 5 pins:

| Pin | Function |
|-|-|
| `VCC` | power source (3V3) |
| `GND` | ground |
| `SCL` | `CLK` line |
| `SDA` | `MOSI` line |
| `CSB` | `CS` line |
| `SDO` | `MISO` line |

![bmp280_wiring](images/bmp280_wiring.png)

:::note
The BMP280 can also be interfaced through I2C, using the same pins but with different functions.
:::

> The Raspberry Pi Pico has two usable SPI channels. Each channel has a set SPI control pins.

Reviewer 2: ```suggestion
The Raspberry Pi Pico has two usable SPI channels. Each channel has a set of SPI control pins.
```

I'm not sure what this wants to mean.
